### PR TITLE
Port firebase from cross-android and jsaddle-cbits

### PR DESCRIPTION
**Pull request by 3noch**

### CODE REVIEW

**Comment by 3noch** [comment]
This hasn't been tested yet.

File: src/Android/HaskellActivity.hsc [hunk 1]
```diff
@@ -118,9 +122,14 @@ activityCallbacksToPtrs ac = ActivityCallbacksPtrs
...
   <*> wrapCStringCStringIO (\a b -> do
-                               a' <- peekCString a
-                               b' <- peekCString b
-                               _activityCallbacks_onNewIntent ac a' b')
+        a' <- peekCString a
+        b' <- peekCString b
+        _activityCallbacks_onNewIntent ac a' b'
+      )
+  <*> wrapCStringIO (\token -> do
+        token' <- peekCString token
```
Comment (3noch) [inline src/Android/HaskellActivity.hsc:130]: In original code this did `T.decodeUtf8 . packCString`. I'm a bit worried about utf8 here.

File: cbits/HaskellActivity.c [hunk 2]
```diff
@@ -86,6 +86,26 @@ JNIEXPORT void JNICALL Java_systems_obsidian_HaskellActivity_haskellOnNewIntent
...
+JNIEXPORT void JNICALL Java_systems_obsidian_LocalFirebaseInstanceIDService_handleDeviceToken (JNIEnv *env, jobject thisObj, jlong callbacksLong, jstring token) {
+  const ActivityCallbacks *callbacks = (const ActivityCallbacks *)callbacksLong;
+  if(callbacks->firebaseInstanceIdService_sendRegistrationToServer) {
+    const char *cstring_token = (*env)->GetStringUTFChars(env, token, 0);
+    callbacks->firebaseInstanceIdService_sendRegistrationToServer(cstring_token);
+    (*env)->ReleaseStringUTFChars(env, token, cstring_token);
+  }
+}
+
+JNIEXPORT void JNICALL Java_systems_obsidian_LocalFirebaseMessagingService_handleNotification (JNIEnv *env, jobject thisObj, jlong callbacksLong, jstring intent, jstring notificationdata) {
```
Comment (3noch) [inline cbits/HaskellActivity.c]: I suspect this method isn't being called in such a way that callbacksLong is supplied.

Comment (3noch) [reply cbits/HaskellActivity.c]: The previous code used a global to store callbacksLong. We could do the same here

Comment (3noch) [reply cbits/HaskellActivity.c]: See #6

**Comment by fendor** [comment]
Any updates on this? 
Apparently this branch is being used in reflex-platform since at least two months

**Pull request merged**